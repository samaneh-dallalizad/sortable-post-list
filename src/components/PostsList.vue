<template>
  <li v-for="(post, index) in posts" :key="post.id" :ref="'post-' + index">
    <span>Post {{ post.id }}</span>
    <a
      v-if="index > 0"
      :class="(index === posts.length - 1 ? 'top-50%' : 'top-0')"
      @click="animateChangeOrder('up', index)"
    >
      <ChevronUpIcon class="h-5 w-5 "/>
    </a>
    <a
      v-if="index < posts.length - 1"
      :class="(index ===  0 ? 'bottom-50%' : 'bottom-0')"
      @click="animateChangeOrder( 'down', index)"
    >
      <ChevronDownIcon class="h-5 w-5"/>
    </a>
  </li>
</template>

<script>
import {ChevronDownIcon, ChevronUpIcon} from '@heroicons/vue/solid'

export default {
  name: "PostsList",
  props: {
    posts: Array,
  },
  emits: [
    "changeOrder"
  ],
  components: {
    ChevronDownIcon,
    ChevronUpIcon
  },
  methods: {
    sleep(ms) {
      return new Promise(resolve => setTimeout(resolve, ms));
    },

    async animateChangeOrder(direction, index) {
      let upElement = null
      let downElement = null
      switch (direction) {
        case 'down':
          downElement = this.$refs["post-" + index][0]
          upElement = this.$refs["post-" + (index + 1)][0]
          break
        default: // up
          upElement = this.$refs["post-" + index][0]
          downElement = this.$refs["post-" + (index - 1)][0]
          break
      }
      upElement.classList.add('slideup')
      downElement.classList.add('slidedown')
      await this.sleep(1000)
      upElement.classList.remove('slideup')
      downElement.classList.remove('slidedown')
      this.$emit('changeOrder', direction, index)
    }
  }

}
</script>

<style scoped>
ul li {
  @apply flex bg-white text-white py-4 px-2 text-[#bebfbf] rounded-md shadow-md relative;
}

ul li a {
  @apply text-[#8076bf] absolute right-0 p-1;
}

.slidedown {
  animation: slidedown 1s;
}

.slideup {
  animation: slideup 1s;
}

@keyframes slidedown {
  from {
    transform: translateY(0%);
  }

  to {
    transform: translateY(128%);
  }
}

@keyframes slideup {
  from {
    transform: translateY(0%);
  }

  to {
    transform: translateY(-128%);
  }
}

</style>