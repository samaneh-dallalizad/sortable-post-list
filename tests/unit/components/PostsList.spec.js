import {shallowMount} from '@vue/test-utils'
import PostsList from "@/components/PostsList";

describe('PostsList.vue', () => {

    it('display 5 posts', () => {
        const wrapper = shallowMount(PostsList, {
            props: {posts: [{id: 1}, {id: 2}, {id: 3}, {id: 4}, {id: 5}]}
        })
        expect(wrapper.element).toMatchSnapshot()
    })

})