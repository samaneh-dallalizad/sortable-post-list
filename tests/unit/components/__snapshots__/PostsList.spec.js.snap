// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsList.vue display 5 posts 1`] = `
<div
  data-v-app=""
>
  
  <li>
    <span>
      Post 1
    </span>
    <!--v-if-->
    <a
      class="bottom-50%"
    >
      <chevron-down-icon-stub
        class="h-5 w-5"
      />
    </a>
  </li>
  <li>
    <span>
      Post 2
    </span>
    <a
      class="top-0"
    >
      <chevron-up-icon-stub
        class="h-5 w-5"
      />
    </a>
    <a
      class="bottom-0"
    >
      <chevron-down-icon-stub
        class="h-5 w-5"
      />
    </a>
  </li>
  <li>
    <span>
      Post 3
    </span>
    <a
      class="top-0"
    >
      <chevron-up-icon-stub
        class="h-5 w-5"
      />
    </a>
    <a
      class="bottom-0"
    >
      <chevron-down-icon-stub
        class="h-5 w-5"
      />
    </a>
  </li>
  <li>
    <span>
      Post 4
    </span>
    <a
      class="top-0"
    >
      <chevron-up-icon-stub
        class="h-5 w-5"
      />
    </a>
    <a
      class="bottom-0"
    >
      <chevron-down-icon-stub
        class="h-5 w-5"
      />
    </a>
  </li>
  <li>
    <span>
      Post 5
    </span>
    <a
      class="top-50%"
    >
      <chevron-up-icon-stub
        class="h-5 w-5"
      />
    </a>
    <!--v-if-->
  </li>
  
</div>
`;
